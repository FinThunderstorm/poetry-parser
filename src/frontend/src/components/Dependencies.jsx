import React from 'react'
import {
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow,
    Typography,
} from '@mui/material'
import { Link } from 'react-router-dom'

/** Dependencies converts dependencies object into table
 */
const Dependencies = ({ dependencies }) => (
    <>
        <Typography variant="button" color="text.secondary">
            Dependencies:
        </Typography>
        <TableContainer sx={{ height: '60vh' }}>
            <Table stickyHeader size="small">
                <TableHead>
                    <TableRow key="head-row">
                        <TableCell>Name</TableCell>
                        <TableCell>Required</TableCell>
                        <TableCell>Installed</TableCell>
                        <TableCell>Source</TableCell>
                        <TableCell>Category</TableCell>
                    </TableRow>
                </TableHead>
                <TableBody>
                    {Object.values(dependencies)
                        .sort((a, b) => {
                            const nameA = a.name.toLowerCase()
                            const nameB = b.name.toLowerCase()
                            if (nameA < nameB) {
                                return -1
                            }
                            if (nameA > nameB) {
                                return 1
                            }
                            return 0
                        })
                        .map((dependency) => (
                            <TableRow key={`${dependency.name}-row`}>
                                {dependency.installed ? (
                                    <TableCell>
                                        <Link
                                            to={`/packages/${dependency.name}`}
                                            title={`dependency-${dependency.name}`}
                                        >
                                            {dependency.name}
                                        </Link>
                                    </TableCell>
                                ) : (
                                    <TableCell>{dependency.name}</TableCell>
                                )}

                                <TableCell>
                                    {dependency.optional ? '❌' : '✅'}
                                </TableCell>

                                <TableCell>
                                    {dependency.installed ? '✅' : '❌'}
                                </TableCell>
                                <TableCell>{dependency.source}</TableCell>
                                <TableCell>{dependency.category}</TableCell>
                            </TableRow>
                        ))}
                </TableBody>
            </Table>
        </TableContainer>
    </>
)

export default Dependencies
