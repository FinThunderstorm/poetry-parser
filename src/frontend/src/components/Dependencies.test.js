/* eslint-disable react/prop-types */
import React from 'react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import { render, screen } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'

import Dependencies from './Dependencies'

describe('Dependencies', () => {
    const DependenciesWrapper = ({ url, dependencies }) => (
        <MemoryRouter initialEntries={[url]}>
            <Routes>
                <Route path="/">
                    <Route path="packages">
                        <Route
                            path=":name"
                            element={
                                <Dependencies dependencies={dependencies} />
                            }
                        />
                    </Route>
                </Route>
            </Routes>
        </MemoryRouter>
    )

    test('Dependencies renderes correctly', () => {
        const dependencies = {
            lockfile: {
                name: 'lockfile',
                category: 'filecache',
                optional: true,
                source: 'extras',
                installed: true,
            },
            msgpack: {
                name: 'msgpack',
                optional: false,
                source: 'dependencies',
                installed: true,
            },
            requests: {
                name: 'requests',
                optional: false,
                source: 'dependencies',
                installed: true,
            },
            redis: {
                name: 'redis',
                category: 'redis',
                optional: true,
                source: 'extras',
                installed: false,
            },
        }
        render(
            <DependenciesWrapper
                url="/packages/cachecontrol"
                dependencies={dependencies}
            />
        )

        const title = screen.getByText('Dependencies:', {
            exact: false,
        })
        const name = screen.getByText('Name', {
            exact: false,
        })
        const required = screen.getByText('Required', {
            exact: false,
        })
        const installed = screen.getByText('Installed', {
            exact: false,
        })
        const source = screen.getByText('Source', {
            exact: false,
        })
        const category = screen.getByText('Category', {
            exact: false,
        })
        expect(title).toBeDefined()
        expect(name).toBeDefined()
        expect(required).toBeDefined()
        expect(installed).toBeDefined()
        expect(source).toBeDefined()
        expect(category).toBeDefined()

        const lockfile = screen.getByText('lockfile', {
            exact: false,
        })
        const filecache = screen.getByText('filecache', {
            exact: false,
        })

        expect(lockfile).toBeDefined()
        expect(filecache).toBeDefined()

        const extras = screen.queryAllByText('extras', {
            exact: false,
        })
        expect(extras.length).toBe(2)

        const check = screen.queryAllByText('✅', {
            exact: false,
        })
        expect(check.length).toBe(5)
        const notCheck = screen.queryAllByText('❌', {
            exact: false,
        })
        expect(notCheck.length).toBe(3)

        const tree = renderer
            .create(
                <DependenciesWrapper
                    url="/packages/cachecontrol"
                    dependencies={dependencies}
                />
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
