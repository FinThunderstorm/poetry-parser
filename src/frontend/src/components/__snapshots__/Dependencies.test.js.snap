// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dependencies Dependencies renderes correctly 1`] = `
Array [
  <span
    className="MuiTypography-root MuiTypography-button css-1jm68wq-MuiTypography-root"
  >
    Dependencies:
  </span>,
  <div
    className="MuiTableContainer-root css-e2p4vx-MuiTableContainer-root"
  >
    <table
      className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
            scope="col"
          >
            Required
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
            scope="col"
          >
            Installed
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
            scope="col"
          >
            Source
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
            scope="col"
          >
            Category
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            <a
              href="/packages/lockfile"
              onClick={[Function]}
              title="dependency-lockfile"
            >
              lockfile
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ❌
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ✅
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            extras
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            filecache
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            <a
              href="/packages/msgpack"
              onClick={[Function]}
              title="dependency-msgpack"
            >
              msgpack
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ✅
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ✅
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            dependencies
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          />
        </tr>
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            redis
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ❌
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ❌
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            extras
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            redis
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            <a
              href="/packages/requests"
              onClick={[Function]}
              title="dependency-requests"
            >
              requests
            </a>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ✅
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            ✅
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            dependencies
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          />
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
